webcompat/static/js/lib/models/query-params.js

```diff
@@ -0,0 +1,360 @@
+var issueList = issueList || {};
+
+/*
+* QueryParams is a model for keeping track of all parameters
+* we need to fetch the list of issues.
+*
+* Dropdown, filter, search and sorting views will update the params,
+* typically by calling
+* this.mainView.params.setParam(name, value).
+* The QueryParams model will observe itself and trigger relevant
+* updates to the various views through firing events.
+*
+* The model can serialize its params to generate URLs. Queries are
+* either proxied through webcompat.com or go directly to GitHub.
+*
+* If the request is a simple query (no search terms / ?q= parameter),
+* it is always proxied through webcompat and the backend uses GitHub's
+* Issues API - https://developer.github.com/v3/issues/
+* Generated URLs are for example
+* /api/issues?page=1&per_page=50&sort=created&direction=desc
+* /api/issues/needsdiagnosis?page=1&per_page=50&sort=created&direction=desc
+*
+* If the request has a search query, and the user is logged in,
+* it is also proxied through webcompat and the backend uses GitHub's
+* Search API - https://developer.github.com/v3/search/#search-issues
+*
+* If the request has a search query and the user is *not* logged in,
+* we send the query directly to GitHub's CORS-enabled Search API.
+* Generated URLs are for example
+* https://api.github.com/search/issues?q=foo&sort=created&order=desc
+*
+* The query itself can contain name:value parts like label:state-needsdiagnosis
+* but we have no control over paging and number of results when querying
+* GitHub directly.
+*/
+issueList.QueryParams = Backbone.Model.extend({
+  defaults:{
+    stage:'all',
+    state: 'open',
+    page:1,
+    per_page:50,
+    sort:'created',
+    direction: 'desc',
+    q: '',
+    creator: '',
+    mentioned: '',
+    labels: [] // Using 'labels' plural form would be nice, but needs to be
+              // singular for the backend - or we'd have to translate name on send
+  },
+  configUrls: {
+    _githubSearch: 'https://api.github.com/search/issues'
+  },
+  bugstatuses: ['contactready', 'needsdiagnosis', 'needscontact', 'sitewait'],
+  initialize: function(){
+    var self = this;
+    this.on('change', function(e){
+      // When the query/parameters change, we want to
+      // * update the drop-down menu if required
+      // * update the URL and push a history entry
+      // * update the relevant views (list of issues, search box)
+      // To do so, this method will fire various events.
+      // However, first we want to check if the change is significant..
+      var significant = false;
+      var changelist = Object.keys(e.changed);
+      for(var i=0, change; change = changelist[i]; i++) {
+        // I think we only get notified of one property at a time so looping is
+        // maybe not necessary here. However..
+        var newvalue = e.changed[change];
+        var oldvalue = e._previousAttributes[change];
```

Comment:
I guess `e.previousAttributes` is an undocumented internal API? Fingers crossed it doesn't disappear :p